DataReader/src/main/java/org/endeavourhealth/pipeline/inbound/helpers/DBConnectionManager.java

```diff
@@ -55,27 +45,87 @@ private static PreparedStatement prepareEventUpsert() throws SQLException {
     return getEventConnection().prepareStatement("INSERT INTO event (id, json) VALUES (?,?) ON CONFLICT (id) DO UPDATE SET json=?::json");
   }
 
-  private static PreparedStatement prepareInstanceUpsert() throws SQLException {
-    return getInstanceConnection().prepareStatement("INSERT INTO instance (id, json) VALUES (?,?) ON CONFLICT (id) DO UPDATE SET json=?::json");
+  private static PreparedStatement prepareInstanceUpsert(String datatype) throws SQLException {
+    if (!datatype.matches("^[a-zA-Z_][a-zA-Z0-9_]*$"))  // check for SQL injection
+      throw new IllegalArgumentException("Invalid table name: " + datatype);
+
+    return getInstanceConnection().prepareStatement("INSERT INTO datatype (id, json) VALUES (?,?) ON CONFLICT (id) DO UPDATE SET json=?::json".replace("datatype", datatype));
+  }
+
+  private static PreparedStatement prepareInstanceCreateTable(String datatype) throws SQLException {
+    if (!datatype.matches("^[a-zA-Z_][a-zA-Z0-9_]*$"))  // check for SQL injection
+      throw new IllegalArgumentException("Invalid table name: " + datatype);
+
+    return getInstanceConnection().prepareStatement("""
+      CREATE OR REPLACE FUNCTION json_date(text)
+        RETURNS timestamptz AS
+      $$SELECT to_timestamp($1, 'YYYY-MM-DDTHH24:MI:SS')$$
+        LANGUAGE sql IMMUTABLE;
+      
+      
+      CREATE TABLE datatype (
+            id UUID PRIMARY KEY,
+            json JSON NOT NULL,
+            type text generated always as (json ->> '@type') stored
+            );
+      
+            CREATE INDEX idx_datatype_pat_dob ON datatype ((json_date(json ->> 'dateOfBirth'))) WHERE type = 'Patient';
+            CREATE INDEX idx_datatype_pat_nhs ON datatype ((json ->> 'nhsNumber')) WHERE type = 'Patient';
+      """.replaceAll("datatype", datatype));
```

Comment:
Instead of `datatype`, can we use `${datatype}` instead? Just makes it more obvious that its a variable/placeholder in the string.